### PR TITLE
[no squash] Various minor fixes

### LINKED ISSUE #10238: Warning -Warray-bounds on H_ns in src/util/srp.cpp
##### Minetest version

Commit:  3ce03d1c2a63d261c83f5962cd13212697f19472
Type: Build issue / warning

##### OS / Hardware

Operating system: Ubuntu 18.04
Compiler: Gcc/G++ (Ubuntu 10.1.0-2ubuntu1~18.04) 10.1.0

##### Summary

Got this compilation warning on build with gcc 10.

##### Steps to reproduce

Compile with gcc 10.

```
[ 31%] Building CXX object src/CMakeFiles/minetest.dir/version.cpp.o
In file included from /usr/include/string.h:494,
                 from /usr/include/c++/10/cstring:42,
                 from minetest/src/util/srp.cpp:43:
In function ‘void* memcpy(void*, const void*, size_t)’,
    inlined from ‘int calculate_x(__mpz_struct*, SRP_HashAlgorithm, const unsigned char*, size_t, const char*, const unsigned char*, size_t)’ at minetest/src/util/srp.cpp:432:8:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:34:33: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ forming offset [64, 4294967294] is out of the bounds [0, 64] of object ‘ucp_hash’ with type ‘unsigned char [64]’ [-Warray-bounds]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
minetest/src/util/srp.cpp: In function ‘int calculate_x(__mpz_struct*, SRP_HashAlgorithm, const unsigned char*, size_t, const char*, const unsigned char*, size_t)’:
minetest/src/util/srp.cpp:443:16: note: ‘ucp_hash’ declared here
  443 |  unsigned char ucp_hash[SHA512_DIGEST_LENGTH];
      |                ^~~~~~~~
```

## Link to code:

https://github.com/minetest/minetest/blob/master/src/util/srp.cpp#L424-L437

## Previous bug on this line

https://github.com/minetest/minetest/pull/6855

### PR DESCRIPTION
**Pull request by ShadowNinja**
* Fix all<sup>[1]</sup> compiler warnings.
  * Fixes #10238.
  * We may want to enable `-Werror`, either locally or on CI to avoid introducing any new warnings, let me know your thoughts.
  * Note that the VoxelArea test changes compile to the same code, but this may be surprising.  We might want to actually change the tests so that they're not checking behavior on overflow.
  * clang 3.9's `-Wmissing-braces` implementation seems to warn in a number of cases where other compiler don't so maybe we want to just disable that warning on the compiler and version.
  * Irrlicht still has a ton of warnings, so if you build Irrlicht as part of the Minetest build (in lib/irrlichtmt) rather than as a separate build you will still get plenty of warnings.
* Use `build` as the build directory on CI.  Previously `cmakebuild` was used because the `build` directory was used for the Android sources.
* Fix `XDG_CACHE_HOME` spec compliance.  The basedir spec says that an empty value should be treated like a missing value.
* Some spacing fixes.
* Auto-detect build parallelism on CI.
  * Not sure how many processors CI has, but this probably shouldn't be hardcoded.  Maybe CI has 1 core and 2 jobs gave a bit more performance?  It could be changed to something like `$(($(nproc) + 1))` if so.
* Add some additional names that are reserved on Windows but undocumented: https://googleprojectzero.blogspot.com/2016/02/the-definitive-guide-on-win32-to-nt.html
* Some test fixes

[1] Except for the clang 3.9 warnings that might be bogus...

### CODE REVIEW

File: .github/workflows/macos.yml [hunk 1]
```diff
@@ -45,8 +45,8 @@ jobs:
           git clone -b $MINETEST_GAME_BRANCH $MINETEST_GAME_REPO games/$MINETEST_GAME_NAME
           rm -rvf games/$MINETEST_GAME_NAME/.git
           git clone https://github.com/minetest/irrlicht -b $IRRLICHT_TAG lib/irrlichtmt
-          mkdir cmakebuild
-          cd cmakebuild
+          mkdir build
+          cd build
           cmake .. \
```
Comment (sfan5) [inline .github/workflows/macos.yml:50]: If the installed cmake isn't ancient you can actually shorten this to `cmake -S .. -B build`
(and then `make -C build` obviously)

File: src/unittest/test_voxelarea.cpp [hunk 2]
```diff
@@ -120,7 +120,7 @@ void TestVoxelArea::test_extent()
 	VoxelArea v1(v3s16(-1337, -547, -789), v3s16(-147, 447, 669));
 	UASSERT(v1.getExtent() == v3s16(1191, 995, 1459));
 
-	VoxelArea v2(v3s16(32493, -32507, 32753), v3s16(32508, -32492, 32768));
+	VoxelArea v2(v3s16(32493, -32507, 32753), v3s16(32508, -32492, -32768));
```
Comment (sfan5) [inline src/unittest/test_voxelarea.cpp:123]: If you do this you're actually testing for an overflow, I suggest instead subtracting 1 from every component to keep it in-range.

File: src/util/string.cpp [hunk 3]
```diff
@@ -846,13 +851,33 @@ static const std::array<std::wstring, 22> disallowed_dir_names = {
...
 
+
+/**
+ * @param str
+ * @return A copy of \p str with trailing whitespace removed.
+ */
+static std::wstring wrtrim(const std::wstring &str)
+{
+	size_t back = str.size();
+	while (back > 0 && std::isspace(str[back - 1]))
```
Comment (sfan5) [inline src/util/string.cpp:874]: doesn't a string have these methods?
```suggestion
	while (!str.empty() && std::isspace(str.back())))
```

File: util/ci/build.sh [hunk 4]
```diff
@@ -1,8 +1,8 @@
...
 
-mkdir cmakebuild
-cd cmakebuild
+mkdir build
+cd build
 cmake -DCMAKE_BUILD_TYPE=Debug \
 	-DRUN_IN_PLACE=TRUE -DENABLE_GETTEXT=TRUE \
 	-DBUILD_SERVER=TRUE ${CMAKE_FLAGS} ..
-make -j2
+make -j$(nproc)
```
Comment (sfan5) [inline util/ci/build.sh:8]: Regarding this I believe the CI has 1 core but 2 was chosen for speed just like you suspected.

Comment (sfan5) [reply util/ci/build.sh:8]: well this is inconclusive:
Unbenannt

File: src/util/string.cpp (hunk 3 above)
Comment (Desour) [reply src/util/string.cpp:874]: Or:
```cpp
auto p = str.find_last_not_of(L" \f\n\r\t\v");
return p == str.npos ? "" : str.substr(0, ++p);
```
(But it doesn't respect the locale.)

**Comment by Desour** [comment]
Can confirm that it compiles without warnings with gcc 11.1.0 .

**Comment by erlehmann** [comment]
> We might want to actually change the tests so that they're not checking behavior on overflow.

That seems *weird* to me. Why would you do that?

File: src/util/string.cpp (hunk 3 above)
Comment (SmallJoker) [reply src/util/string.cpp:874]: No need to re-invent the wheel. `std::isspace` exists to avoid such hardcoded strings.

File: minetest.conf.example [hunk 5]
```diff
@@ -1385,7 +1385,7 @@
 # ask_reconnect_on_crash = false
 
 #    From how far clients know about objects, stated in mapblocks (16 nodes).
-#    
+#
```
Comment (SmallJoker) [inline minetest.conf.example:1388]: Wrong file. `settingtypes.txt` should be fixed and then exported to `minetest.conf.example` with this code:
https://github.com/minetest/minetest/blob/9d3135a21b31eea6dcec67b887243bcb6124b6a1/builtin/mainmenu/dlg_settings_advanced.lua#L1105-L1110

Comment (sfan5) [reply minetest.conf.example:1388]: Or rather the code be fixed, it generates these trailing spaces.

File: src/util/string.cpp (hunk 3 above)
Comment (ShadowNinja) [reply src/util/string.cpp:874]: `back` is changing on each iteration and this code doesn't modify `str` (it's immutable since it's passed as a `const&`), so we can't just use `str.empty()` and `str.back()`.

I'm going to change this to something entirely different though:  leading and trailing ASCII spaces will be replaced with underscores since those cause problems but not other kinds of whitespace according to this: https://support.microsoft.com/en-us/topic/d9813c8e-81f3-ac34-f82e-7db83ade92c2

`\f\n\r\t\v` and other control characters are already blocked.

File: builtin/mainmenu/generate_from_settingtypes.lua [hunk 6]
```diff
@@ -31,7 +31,7 @@ local group_format_template = [[
 #    octaves     = %s,
 #    persistence = %s,
 #    lacunarity  = %s,
-#    flags       = %s
+#    flags       =%s
```
Comment (ShadowNinja) [inline builtin/mainmenu/generate_from_settingtypes.lua:34]: This is a bit ugly, but I can't think of a better way to do it that doesn't involve more substantial changes.

**Comment by rubenwardy** [comment]
I think it would be more accurate to replace "Various" with "A metric shit tonne" 

Thanks, this is good to see

**Comment by ShadowNinja** [comment]
Pushed some changes based on comments.  I intend to squash the changes into their respective original fix commits before this is merged.

I also found a typo and fixed that.

I also started using the `--build` and `--install` CMake options instead of calling the build tool directly.  This is better because it ensures that the build is conducted with the same version of the tool that CMake detected during the configure step.

File: src/settings_translation_file.cpp [hunk 7]
```diff
@@ -263,29 +263,6 @@ fake_function() {
...
-	gettext("Shadow filter quality");
-	gettext("Define shadow filtering quality.\nThis simulates the soft shadows effect by applying a PCF or Poisson disk\nbut also uses more resources.");
-	gettext("Colored shadows");
-	gettext("Enable colored shadows.\nOn true translucent nodes cast colored shadows. This is expensive.");
-	gettext("Map shadows update frames");
-	gettext("Spread a complete update of shadow map over given amount of frames.\nHigher values might make shadows laggy, lower values\nwill consume more resources.\nMinimum value: 1; maximum value: 16");
-	gettext("Soft shadow radius");
-	gettext("Set the soft shadow radius size.\nLower values mean sharper shadows, bigger values mean softer shadows.\nMinimum value: 1.0; maximum value: 10.0");
-	gettext("Sky Body Orbit Tilt");
-	gettext("Set the tilt of Sun/Moon orbit in degrees.\nValue of 0 means no tilt / vertical orbit.\nMinimum value: 0.0; maximum value: 60.0");
```
Comment (sfan5) [inline src/settings_translation_file.cpp:288]: Thse should not be removed

File: src/unittest/test_voxelarea.cpp (hunk 2 above)
Comment (sfan5) [reply src/unittest/test_voxelarea.cpp:123]:
```suggestion
	VoxelArea v2(v3s16(32493, -32507, 32752), v3s16(32508, -32492, 32767));
```
likewise below

File: src/util/srp.cpp [hunk 8]
```diff
@@ -422,8 +422,13 @@ static SRP_Result H_nn(
 }
 
 static SRP_Result H_ns(mpz_t result, SRP_HashAlgorithm alg, const unsigned char *n,
-	size_t len_n, const unsigned char *bytes, uint32_t len_bytes)
+	size_t len_n, const unsigned char *bytes, size_t len_bytes)
 {
+#ifdef __GNUC__
+	// Silence warning
+	if (len_bytes >= PTRDIFF_MAX)
```
Comment (sfan5) [inline src/util/srp.cpp:429]: I think changing this line would silence the warning as well:
https://github.com/minetest/minetest/blob/22b3a5585e48e0822eb97b1b9437b9af80763320/src/util/srp.cpp#L357

File: src/settings_translation_file.cpp (hunk 7 above)
Comment (ShadowNinja) [reply src/settings_translation_file.cpp:288]: This was updated by `generate_from_settingtypes.lua`.  These strings don't seem to actually be used.

File: src/terminal_chat_console.cpp [hunk 9]
```diff
@@ -26,6 +26,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "util/numeric.h"
 #include "util/string.h"
 #include "chat_interface.h"
+#include <inttypes.h>
```
[the remark this answers is not in the document]
Comment (ShadowNinja) [reply src/terminal_chat_console.cpp:29]: I didn't find a formal guideline, but that does seem to be the convention, I'll fix it.

File: src/util/srp.cpp (hunk 8 above)
Comment (ShadowNinja) [reply src/util/srp.cpp:429]: Ah, that seems to be the real source of the issue.  I'll change it to 0.

File: src/settings_translation_file.cpp (hunk 7 above)
Comment (ShadowNinja) [reply src/settings_translation_file.cpp]: Nevermind, they're been re-added.  I rebased and re-ran the generator.

File: src/util/string.cpp [hunk 10]
```diff
@@ -870,7 +875,21 @@ std::string sanitizeDirName(const std::string &str, const std::string &optional_
...
+	size_t end = safe_name.find_last_not_of(L' ');
+	if (start == std::wstring::npos || end == std::wstring::npos) {
+		start = end = safe_name.size();
+	}
+	for (size_t i = 0; i < start; i++) {
+		safe_name[i] = L'_';
+	}
+	for (size_t i = end + 1; i < safe_name.size(); i++) {
+		safe_name[i] = L'_';
+	}
```
[the remark this answers is not in the document]
Comment (ShadowNinja) [reply src/util/string.cpp]: Done.

**Comment by ShadowNinja** [comment]
Rebased and addressed comments.

**Review by sfan5 — APPROVED** [review]
LGTM but this comment still applies -> https://github.com/minetest/minetest/pull/12015/files#r806089154

File: src/unittest/test_voxelarea.cpp (hunk 2 above)
Comment (ShadowNinja) [reply src/unittest/test_voxelarea.cpp:123]: This was already checking overflow behavior before, so I left it so as to not change behavior, but that was probably not intended.  I'll make that change.
